src/stan/model/indexing/rvalue.hpp

```diff
@@ -23,297 +48,673 @@ namespace model {
+ *  Eigen matrix.
+ *
+ * Types:  mat[multi, multi] : mat
+ *
+ * @tparam EigMat An eigen matrix
+ * @param[in] x Matrix to index.
+ * @param[in] idxs Pair of multiple indexes.
  * @param[in] name String form of expression being evaluated.
  * @param[in] depth Depth of indexing dimension.
  * @return Result of indexing matrix.
  */
-template <typename T, typename I>
-inline std::enable_if_t<!std::is_same<I, index_uni>::value,
-                        Eigen::Matrix<T, 1, Eigen::Dynamic> >
-rvalue(
-    const Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>& a,
-    const cons_index_list<index_uni, cons_index_list<I, nil_index_list> >& idx,
+template <typename EigMat,
+          stan::internal::require_eigen_dense_dynamic_t<EigMat>* = nullptr>
+inline plain_type_t<EigMat> rvalue(
+    EigMat&& x,
+    const cons_index_list<index_multi,
+                          cons_index_list<index_multi, nil_index_list>>& idxs,
     const char* name = "ANON", int depth = 0) {
-  int m = idx.head_.n_;
-  math::check_range("matrix[uni,multi] indexing, row", name, a.rows(), m);
-  Eigen::Matrix<T, 1, Eigen::Dynamic> r = a.row(m - 1);
-  return rvalue(r, idx.tail_);
+  const auto& x_ref = stan::math::to_ref(x);
+  const int rows = idxs.head_.ns_.size();
+  const int cols = idxs.tail_.head_.ns_.size();
+  plain_type_t<EigMat> x_ret(rows, cols);
+  for (int j = 0; j < cols; ++j) {
+    for (int i = 0; i < rows; ++i) {
+      const int m = idxs.head_.ns_[i];
+      const int n = idxs.tail_.head_.ns_[j];
+      math::check_range("matrix[multi,multi] row index", name, x_ref.rows(), m);
+      math::check_range("matrix[multi,multi] col index", name, x_ref.cols(), n);
+      x_ret.coeffRef(i, j) = x_ref.coeff(m - 1, n - 1);
+    }
+  }
+  return x_ret;
 }
 
 /**
- * Return the result of indexing the specified Eigen matrix with a
- * sequence consisting of a multiple index and a single index,
- * returning a vector.
+ * Return a column of an Eigen matrix with the range of the column specificed
+ *  by another index.
  *
- * Types:  mat[multiple,single] : vector
+ * Types:  mat[Idx, uni] : vec
  *
- * @tparam T Scalar type.
- * @tparam I Type of multiple index.
- * @param[in] a Matrix to index.
- * @param[in] idx Pair multiple index and single index.
+ * @tparam EigMat Eigen type with dynamic rows and columns.
+ * @param[in] x Eigen matrix.
+ * @param[in] idxs Index consisting of one uni-index.
+ * @param[in] name String form of expression being evaluated.
+ * @param[in] depth Depth of indexing dimension.
+ */
+template <typename EigMat, typename Idx,
+          stan::internal::require_eigen_dense_dynamic_t<EigMat>* = nullptr>
+inline auto rvalue(
+    EigMat&& x,
+    const cons_index_list<Idx, cons_index_list<index_uni, nil_index_list>>&
+        idxs,
+    const char* name = "ANON", int depth = 0) {
+  math::check_range("matrix[..., uni] indexing", name, x.cols(),
+                    idxs.tail_.head_.n_);
+  return deep_copy(rvalue(x.col(idxs.tail_.head_.n_ - 1),
```

Comment:
Why `deep_copy`?

Reply:
Ah this was a defensive thing, I was worried that the first index could be a `nil_index_list`, but looking at a bunch of generated stan code idt that's possible so these can be removed

Reply:
Actually there was an edge case this was catching but I think I solved it better. In the case of something like 

```
matrix[5, 5] A = B[, 1:5]
matrix[5, 5] A = B[, 5:1]
```

That goes through the column `index_min_max` specialization that does

```cpp
  if (idxs.tail_.head_.is_ascending()) {
    const auto col_start = idxs.tail_.head_.min_ - 1;
    return rvalue(x.middleCols(col_start, idxs.tail_.head_.max_ - col_start),
                  index_list(idxs.head_), name, depth + 1);
  } else {
    const auto col_start = idxs.tail_.head_.max_ - 1;
    return rvalue(x.middleCols(col_start, idxs.tail_.head_.min_ - col_start)
                      .rowwise()
                      .reverse(),
                  index_list(idxs.head_), name, depth + 1);
  }
```

which in the case of the first index (`idxs.head_`) being `index_omni` the `auto` in the return type here will get confused because one is a `Block<>` and the other is a `Reverse<Block<>>`. I just added a specialization for `index_omni()` for a single index that eval's expressions and passes through plain types.